modules/data.land/R/extract_soil_nc.R

```diff
@@ -30,172 +24,127 @@
-    resolution = grid_spacing, crs = proj_crs$wkt
-  )
-  grid_coords <- terra::crds(raster_template)
+  mukey_values <- terra::values(mu_raster)
+  mukey_values <- mukey_values[!is.na(mukey_values)]
+  mukey_counts <- table(mukey_values)
+  mukeys_all <- as.character(names(mukey_counts))
   
-  # Transform grid coordinates back to WGS84 for gSSURGO queries
-  grid_sf <- sf::st_as_sf(data.frame(x = grid_coords[, 1], y = grid_coords[, 2]),
-                          coords = c("x", "y"), crs = proj_crs)
-  grid_wgs84 <- sf::st_transform(grid_sf, wgs84_crs)
-  grid_coords_wgs84 <- sf::st_coordinates(grid_wgs84)
+  if (length(mukeys_all) == 0) {
+    PEcAn.logger::logger.severe("No mapunit keys were found for this site.")
+  }
   
-  # Query gSSURGO for each grid point to capture spatial variability
-  buffer_radius <- grid_spacing / 2
-  PEcAn.logger::logger.warn(
-    "Buffer radius set to grid_spacing/2 to avoid overlap",
-    "results may be biased due to lack of area weighting and incomplete spatial coverage."
-  )
-  mukeys_all <- c()
-  for (i in seq_len(nrow(grid_coords_wgs84))) {
-    # Extract coordinates for this grid point (not user input)
-    this_lon <- grid_coords_wgs84[i, 1]
-    this_lat <- grid_coords_wgs84[i, 2]
-    
-    # I ask the gSSURGO to find all the mukeys (loosely can be thought of soil type) within grid_spacing distance of each grid point location. 
-    # Basically I think of this as me going around and taking soil samples at each grid point.
-    #https://sdmdataaccess.nrcs.usda.gov/SpatialFilterHelp.htm
-    mu.Path <- paste0(
-      "https://sdmdataaccess.nrcs.usda.gov/Spatial/SDMWGS84Geographic.wfs?",
-      "SERVICE=WFS",
-      "&VERSION=1.1.0",
-      "&REQUEST=GetFeature&TYPENAME=MapunitPoly",
-      "&FILTER=",
-        "<Filter>",
-          "<DWithin>",
-            "<PropertyName>Geometry</PropertyName>",
-            "<gml:Point>",
-              "<gml:coordinates>", this_lon, ",", this_lat, "</gml:coordinates>",
-            "</gml:Point>",
-            "<Distance%20units=%27m%27>", buffer_radius, "</Distance>",
-          "</DWithin>",
-        "</Filter>",
-      "&OUTPUTFORMAT=XMLMukeyList"
+  # Fetch complete soil data
+  sda_data <- tryCatch({
+    soilDB::fetchSDA(
+      WHERE = paste0("mukey IN (", paste(mukeys_all, collapse = ","), ")"),
+      duplicates = TRUE,
+      childs = TRUE,
+      nullFragsAreZero = TRUE,
+      rmHzErrors = TRUE
     )
+  }, error = function(e) {
+    PEcAn.logger::logger.error(paste("Failed to fetch SDA data:", e$message))
+    return(NULL)
+  })
+  
+  if (is.null(sda_data)) {
+    PEcAn.logger::logger.error("Could not retrieve soil data from SDA")
+    return(NULL)
+  }
+  
+  hz_data <- aqp::horizons(sda_data)
+  site_data <- aqp::site(sda_data)
+  
+  # Component-level aggregation by depth
+  depths_cm <- depths * 100
+  all_soil_data <- list()
+  
+  for (i in seq_along(depths_cm)) {
+    if (i == 1) {
+      top_depth <- 0
```

Comment:
FYI you could eliminate this `if` by requiring the first depth to be 0 and then shifting the indexing. This isn't unreasonable as it's the exact same behaviour as `hist` when you set a vector of breaks (i.e. there is one more break than there is bins)